contracts/CPKFactory.sol

```diff
@@ -1,54 +1,53 @@
 pragma solidity >=0.5.0 <0.7.0;
 
-import { Enum } from "@gnosis.pm/safe-contracts/contracts/common/Enum.sol";
-import { Proxy } from "@gnosis.pm/safe-contracts/contracts/proxies/Proxy.sol";
+import { GnosisSafeProxy } from "@gnosis.pm/safe-contracts/contracts/proxies/GnosisSafeProxy.sol";
+import { GnosisSafeProxyFactory } from "@gnosis.pm/safe-contracts/contracts/proxies/GnosisSafeProxyFactory.sol";
 import { GnosisSafe } from "@gnosis.pm/safe-contracts/contracts/GnosisSafe.sol";
 
+import { ProxyImplSetter } from "./ProxyImplSetter.sol";
+
 contract CPKFactory {
-    event ProxyCreation(Proxy proxy);
+    ProxyImplSetter public proxyImplSetter;
+    GnosisSafeProxyFactory public gnosisSafeProxyFactory;
+
+    constructor(GnosisSafeProxyFactory _gnosisSafeProxyFactory) public {
+        proxyImplSetter = new ProxyImplSetter(address(this));
+        gnosisSafeProxyFactory = _gnosisSafeProxyFactory;
+    }
+
+    event ProxyCreation(GnosisSafeProxy proxy);
 
     function proxyCreationCode() external pure returns (bytes memory) {
-        return type(Proxy).creationCode;
+        return type(GnosisSafeProxy).creationCode;
     }
 
     function createProxyAndExecTransaction(
-        address masterCopy,
-        uint256 saltNonce,
+        address owner,
+        address safeVersion,
+        uint256 saltNonceSalt,
         address fallbackHandler,
-        address to,
-        uint256 value,
-        bytes calldata data,
-        Enum.Operation operation
+        bytes calldata execTxCalldata
     )
         external
+        payable
         returns (bool execTransactionSuccess)
     {
-        GnosisSafe proxy;
-        bytes memory deploymentData = abi.encodePacked(type(Proxy).creationCode, abi.encode(masterCopy));
-        bytes32 salt = keccak256(abi.encode(msg.sender, saltNonce));
-        // solium-disable-next-line security/no-inline-assembly
-        assembly {
-            proxy := create2(0x0, add(0x20, deploymentData), mload(deploymentData), salt)
-        }
-        require(address(proxy) != address(0), "create2 call failed");
+        bytes32 saltNonce = keccak256(abi.encode(owner, saltNonceSalt));
+
+        gnosisSafeProxyFactory.createProxyWithNonce(
+            proxyImplSetter,
+            abi.encodeWithSelector(proxyImplSetter.setImplementation.selector, safeVersion),
+            saltNonce
+        );
 
         {
             address[] memory tmp = new address[](1);
-            tmp[0] = address(this);
-            proxy.setup(tmp, 1, address(0), "", fallbackHandler, address(0), 0, address(0));
+            tmp[0] = address(owner);
+            GnosisSafe(proxy).setup(tmp, 1, address(0), "", fallbackHandler, address(0), 0, address(0));
```

Comment:
calling `setup` here assumes, that all future Safe mastercopies will have a `setup` method with the same parameters (which is unlikely, as it is planned to change this for v1.2.0)

Reply:
The way I am resolving this is by generalizing the setup and the execTransaction to an array of low level calls with value. Note that this should also address the issue with how execTransaction with value is handled where it:

> Might be better to do it in 2 separate calls